### PR TITLE
[CVE-2022-25758] Use dart-sass instead of node-sass

### LINKED ISSUE #1842: CVE-2022-25758 (High) detected in scss-tokenizer-0.2.3.tgz - autoclosed
## CVE-2022-25758 - High Severity Vulnerability
<details><summary> Vulnerable Library - <b>scss-tokenizer-0.2.3.tgz</b></p></summary>

<p>A tokenzier for Sass' SCSS syntax</p>
<p>Library home page: <a href="https://registry.npmjs.org/scss-tokenizer/-/scss-tokenizer-0.2.3.tgz">https://registry.npmjs.org/scss-tokenizer/-/scss-tokenizer-0.2.3.tgz</a></p>
<p>

Dependency Hierarchy:
  - @osd/ui-framework-1.0.0.tgz (Root Library)
    - node-sass-6.0.1.tgz
      - sass-graph-2.2.5.tgz
        - :x: **scss-tokenizer-0.2.3.tgz** (Vulnerable Library)
<p>Found in base branch: <b>main</b></p>
</p>
</details>
<p></p>
<details><summary> Vulnerability Details</summary>
<p>  
  
All versions of package scss-tokenizer are vulnerable to Regular Expression Denial of Service (ReDoS) via the loadAnnotation() function, due to the usage of insecure regex.

<p>Publish Date: 2022-07-01
<p>URL: <a href=https://vuln.whitesourcesoftware.com/vulnerability/CVE-2022-25758>CVE-2022-25758</a></p>
</p>
</details>
<p></p>
<details><summary> CVSS 3 Score Details (<b>7.5</b>)</summary>
<p>

Base Score Metrics:
- Exploitability Metrics:
  - Attack Vector: Network
  - Attack Complexity: Low
  - Privileges Required: None
  - User Interaction: None
  - Scope: Unchanged
- Impact Metrics:
  - Confidentiality Impact: None
  - Integrity Impact: None
  - Availability Impact: High
</p>
For more information on CVSS3 Scores, click <a href="https://www.first.org/cvss/calculator/3.0">here</a>.
</p>
</details>
<p></p>
<details><summary> Suggested Fix</summary>
<p>

<p>Type: Upgrade version</p>
<p>Origin: <a href="https://nvd.nist.gov/vuln/detail/CVE-2022-25758">https://nvd.nist.gov/vuln/detail/CVE-2022-25758</a></p>
<p>Release Date: 2022-07-01</p>
<p>Fix Resolution: no_fix</p>

</p>
</details>
<p></p>

### PR DESCRIPTION
**Pull request by Flyingliuhub**
Signed-off-by: Tao liu <liutaoaz@amazon.com>

### Description
This PR fixes the [Regular expression denial of service in scss-tokenizer](https://github.com/opensearch-project/OpenSearch-Dashboards/security/dependabot/62), use dart-sass instead of node-sass.

The node-sass are deprecated, the detail [here](https://www.npmjs.com/package/node-sass). 

The suggested solution (#535) is that use [dart-sass](https://github.com/sass/dart-sass#using-dart-sass) instead of node-sass

The scan detail as following and link [here](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1842) 

```
yarn why scss-tokenizer
yarn why v1.22.18
[1/4] Why do we have the module "scss-tokenizer"...?
[2/4] Initialising dependency graph...
warning Resolution field "typescript@4.0.2" is incompatible with requested version "typescript@~4.5.2"
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "scss-tokenizer@0.2.3"
info Reasons this module exists
   - "_project_#@osd#ui-framework#node-sass#sass-graph" depends on it
   - Hoisted from "_project_#@osd#ui-framework#node-sass#sass-graph#scss-tokenizer"
info Disk size without dependencies: "268KB"
info Disk size with unique dependencies: "1.11MB"
info Disk size with transitive dependencies: "1.11MB"
info Number of shared dependencies: 2
Done in 1.01s.
```

 
### Issues Resolved
#1842 , #535 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
    - [x] `yarn test:jest`
    - [x] `yarn test:jest_integration`
    - [x] `yarn test:ftr`
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by kavilla** [comment]
looks like a [snapshot failure](https://github.com/opensearch-project/OpenSearch-Dashboards/runs/7640126693?check_suite_focus=true), could update them?

**Comment by kavilla** [comment]
Do we see the issue called out here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/535 with EUI complaining?

I know we bumped EUI so it might have unblocked https://github.com/opensearch-project/OpenSearch-Dashboards/issues/535.

**Comment by Flyingliuhub** [comment]
> EUI complaining?

According the error in the #535, it sounds like that happened on during the build compile, I didn't see the build compile error and UI error.

**Comment by Flyingliuhub** [comment]
> looks like a [snapshot failure](https://github.com/opensearch-project/OpenSearch-Dashboards/runs/7640126693?check_suite_focus=true), could update them?

Thanks. I updated the test, it should fixes the tests.

**Review by seraphjiang — APPROVED** [review]

File: packages/osd-optimizer/src/worker/webpack.config.ts [hunk 1]
```diff
@@ -176,9 +176,8 @@ export function getWebpackConfig(bundle: Bundle, bundleRefs: BundleRefs, worker:
                       )};\n${content}`;
                     },
                     webpackImporter: false,
-                    implementation: require('node-sass'),
+                    implementation: require('sass'),
                     sassOptions: {
-                      outputStyle: 'nested',
```
Comment (kavilla) [inline packages/osd-optimizer/src/worker/webpack.config.ts:181]: is this because the default is `nested` so this is unneeded?

Comment (Flyingliuhub) [reply packages/osd-optimizer/src/worker/webpack.config.ts:181]: dart-sass didn't support nested. https://github.com/sass/dart-sass/issues/442

File: yarn.lock [hunk 2]
```diff
@@ -8868,13 +8858,6 @@ gauge@~2.7.3:
     strip-ansi "^3.0.1"
     wide-align "^1.1.0"
 
-gaze@^1.0.0:
```
Comment (kavilla) [inline yarn.lock:8871]: was this automatic? seems to be added for grunt.

File: yarn.lock [hunk 3]
```diff
@@ -9207,15 +9185,6 @@ globjoin@^0.1.4:
   resolved "https://registry.yarnpkg.com/globjoin/-/globjoin-0.1.4.tgz#2f4494ac8919e3767c5cbb691e9f463324285d43"
   integrity sha1-L0SUrIkZ43Z8XLtpHp9GMyQoXUM=
 
-globule@^1.0.0:
```
Comment (kavilla) [inline yarn.lock:9210]: dependency of the gaze, so whatever the resolution is there will impact this as well.

File: yarn.lock [hunk 4]
```diff
@@ -12073,7 +12037,7 @@ lodash.uniq@^4.5.0:
   resolved "https://registry.yarnpkg.com/lodash.uniq/-/lodash.uniq-4.5.0.tgz#d0225373aeb652adc1bc82e4945339a842754773"
   integrity sha1-0CJTc662Uq3BvILklFM5qEJ1R3M=
 
-lodash@4.17.21, lodash@^4.0.0, lodash@^4.0.1, lodash@^4.10.0, lodash@^4.17.10, lodash@^4.17.11, lodash@^4.17.13, lodash@^4.17.14, lodash@^4.17.15, lodash@^4.17.19, lodash@^4.17.21, lodash@^4.17.4, lodash@^4.7.0, lodash@~4.17.10, lodash@~4.17.15, lodash@~4.17.19, lodash@~4.17.21:
```
Comment (kavilla) [inline yarn.lock:12076]: nit: same as above, i'd would keep these resolutions but i'm overly cautious.

File: yarn.lock (hunk 2 above)
Comment (Flyingliuhub) [reply yarn.lock:8871]: It was added automatically.

**Comment by kavilla** [comment]
Awesome! Do you know if there is any performance updates with the rendering? Also, will OUI be impacted by this @AMoo-Miki ? Finally, I haven't checked by do we use the sass `render` or `renderSync`?

File: packages/osd-optimizer/src/worker/webpack.config.ts (hunk 1 above)
Comment (kavilla) [reply packages/osd-optimizer/src/worker/webpack.config.ts:181]: I see the default for dart-sass is `expanded` is it worth discussing if it should be `compressed`? Pros and cons? Based on the naming of this package as optimization I'd would think `removes as many extra characters as possible` would ideally be the route.

Comment (Flyingliuhub) [reply packages/osd-optimizer/src/worker/webpack.config.ts:181]: I don't see any pros and cons for changing to compressed. will change it to compressed

File: yarn.lock [hunk 5]
```diff
@@ -9047,7 +9025,7 @@ glob@7.1.3:
     once "^1.3.0"
     path-is-absolute "^1.0.0"
 
-glob@7.1.7, glob@~7.1.1, glob@~7.1.6:
```
[the remark this answers is not in the document]
Comment (Flyingliuhub) [reply yarn.lock]: Thanks. fixed

**Comment by Flyingliuhub** [comment]
@AMoo-Miki , Is this changes looking good from your side? if you have time, please take a look this changes, thanks

**Review by AMoo-Miki — APPROVED** [review]

**Comment by Flyingliuhub** [comment]
> Hello @Flyingliuhub, are we able to eyeball any performance impact?

I didn't see the page load (dashboard/discover/Visualizations) different in my local. 

It take about 12 minutes for build locally with following command 

```
yarn build-platform linux --skip-os-packages --skip-archives --release
```

image

**Review by seraphjiang — APPROVED** [review]

**Review by ananzh — APPROVED** [review]
LGTM

**Comment by Flyingliuhub** [comment]
@kavilla,  Please let me know the feedback from retro. If everything good, can you help to merge it as we discussed. Thanks

**Comment by seraphjiang** [comment]
What are possible risk and impact to get this merged? Could we list and discuss here.

Is it a check list we could follow to self check besides automation test

**Comment by kavilla** [comment]
> What are possible risk and impact to get this merged? Could we list and discuss here.
> 
> Is it a check list we could follow to self check besides automation test

Biggest worry was OUI, but Miki is good with that. Downstream might be impacted so it will be a breaking change for 3.0. The only other thing is if there is any performance degradation. But we don't have any client side perf tests so looks good to me. However, will hold off from merging until at least tomorrow just in case in becomes a blocker for OUI.

**Comment by seraphjiang** [comment]
Any update team ~
@Flyingliuhub @kavilla @AMoo-Miki

**Review by kavilla — APPROVED** [review]

**Pull request merged**